code/mission/missioncampaign.cpp

```diff
@@ -1554,14 +1554,47 @@ int mission_load_up_campaign(bool fall_back_from_current)
 	auto pl = Player;
 
 	// find best campaign to use:
-	//   1) last used
-	//   2) builtin
-	//   3) anything else
+	//   1) cmdline
+	//   2) last used
+	//   3) builtin
+	//   4) anything else
 	// Note that in each step we only fall back when the error is benign, e.g. ignored or missing;
 	// if there's some other real error with the campaign file, we report it.
 	// Also note that if we *have* a current campaign, we shouldn't fall back *at all*, lest we repeatedly
 	// reset what the current campaign is, *unless* we are starting a brand new session or loading a new pilot.
 
+	// cmdline...
+	if ( Cmdline_campaign != nullptr && strlen(Cmdline_campaign) ) {
+		char* campaign = Cmdline_campaign;
+
+		// Clear cmdline value
+		// * Only set campaign once from cmdline.
+		// * Prevent subsequent load failures.
+		// * On success, campaign becomes "last used".
+		Cmdline_campaign = nullptr;
+
+		bool has_last_used_campaign = strlen(pl->current_campaign) > 0;
+		bool campaign_already_set = has_last_used_campaign
+			&& (stricmp(campaign, pl->current_campaign) == 0);
+
+		if (has_last_used_campaign) {
+			mprintf(("Current campaign is '%s'\n", pl->current_campaign));
+		}
+
+		if (!campaign_already_set) {
+			rc = mission_campaign_load(campaign, nullptr, pl);
+
+			if (rc == 0) {
+				// update pilot with the new current campaign (becomes "last used")
+				strcpy_s(pl->current_campaign, Campaign.filename);
+				mprintf(("Set current campaign to '%s'\n", campaign));
+				return rc;
+			} else {
+				mprintf(("Failed to set current campaign to '%s'\n", campaign));
+			}
```

Comment:
I'm not 100% satisfied with this code but it's tricky to come up with an alternative.  There are two aspects that don't sit quite right:
1. With this change, there are now three spots in the function that both return `rc` *and* each spot that returns `rc` needs to make sure `pl->current_campaign` is updated on success.
2. If there is an error, the error code is lost.

It would be nice if the code logic simply cascaded from here down to cases 2, 3, and 4, but that would mean modifying the conditions for subsequent checks and it would just be moving the complexity from one spot to another.

Reply:
After further discussion, point 1 is probably fine, and point 2 is not applicable because the `-campaign` option is intended to fall back to existing logic if there's an issue.  So this is good to approve.